misc/python/materialize/cli/ci_flake_detect.py

```diff
@@ -0,0 +1,116 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# ci_flake_detect.py - Detect flaky test failures based on comparing logs to patterns from Github issues
+
+import argparse
+import re
+import sys
+from typing import Any, List
+
+import junit_xml
+import requests
+
+from materialize import ci_util
+
+CI_RE = re.compile("^ci-regexp: (.*)$", flags=re.MULTILINE)
+FAILURE_RE = re.compile("(panicked|Out of memory)")
+
+
+class Flake:
+    def __init__(self, regex: str, info: Any):
+        self.regex = re.compile(regex)
+        self.info = info
+
+
+def main(argv: List[str]) -> int:
+    parser = argparse.ArgumentParser(
+        prog="ci-flake-detect",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description="""
+ci-flake-detect detects known flakes using the log files and open Github issues
+on Materialize repository.""",
+    )
+
+    parser.add_argument(
+        "--name", required=True, help="identifier to use for creating junit xml file"
+    )
+
+    parser.add_argument("log_files", nargs="+", help="log files to search in")
+    args = parser.parse_args()
+
+    annotate_known_failures(args.log_files, args.name)
+    return 0
+
+
+def annotate_known_failures(log_files: List[str], name: str) -> None:
+    """
+    Called when a test failure occurred, check if it is a known flaky CI failure by
+    checking the existing Github issues.
+    """
+
+    failure_lines = get_failure_lines(log_files)
+
+    if not failure_lines:
+        print("No failures found in logs")
+        return
+
+    flakes = get_flakes_from_github()
+
+    junit_suite = junit_xml.TestSuite("Flakes")
+
+    junit_report = ci_util.junit_report_filename(f"{name}_flakes")
+    for i, line in enumerate(failure_lines):
```

Comment:
The line and the file the line came from must be made a part of the information that is being provided about each failure.